Mng 102 finish fight implementation

daria305 opened the pull request.

AgataLudwiczynska committed:
Function RenderChoosePokemonScreen

daria305 committed:
MNG-102 fight method implemented

daria305 committed:
MNG-102 fight method implemented

daria305 committed:
MNG-102 Add tests for fight class. Fix pokemon subtract HP

daria305 committed:
MNG-102 Add tests for fight class. Fix pokemon subtract HP

lukaszdutka commented on src/__tests__/fightClass.spec.ts:
```diff
@@ -1,14 +1,16 @@
-        const attackTypes: string[] = ["normal", "electric", "fire"]
-        const pokemonTypes: string[][] = [["normal"], ["normal", "water"], ["bug", "grass"]]
+        const attackTypes: PokemonType[] = ["normal", "electric", "fire"]
+        const pokemonTypes: PokemonType[][] = [["normal"], ["normal", "water"], ["bug", "grass"]]
```
Why not separate this into three different tests?
I see there three tests:
"When there is no change"
"When there is x2 damage"
"When there is x4 damage"

lukaszdutka replied:
The other way is to create **parameterised tests** here :)
See:
https://medium.com/@suman.maity112/parameterized-tests-with-jest-2932f42b1509

lukaszdutka replied:
The idea is to start with the most simple test: "Two pokemons, one attack another, no modifires/effectiveness" then add more complicated ones: "fire vs water will be 0,5 times damage", "water vs fire will be 2 times damage", "water vs [fire, rock] will be X times damage" etc.

lukaszdutka commented on src/fightClass.ts:
```diff
@@ -22,18 +25,30 @@ export class Fight {
+    private _calcualteDamage(attackingPokemon: Pokemon, deffendingPokemon: Pokemon, move: PokemonMove): number {
+        const minimumDamage = 5;
+        const baseDamage: number = attackingPokemon.attackPoints - deffendingPokemon.defensePoints;
+        return (baseDamage > minimumDamage ? baseDamage : minimumDamage) * this.calculateAttackEffectiveness(move.moveType, deffendingPokemon.types);
```
shorthand:
```suggestion
        return Math.max(baseDamage, minimumDamage) * this.calculateAttackEffectiveness(move.moveType, deffendingPokemon.types);
```

lukaszdutka commented on src/pokemonClass.ts:
```diff
@@ -50,12 +53,18 @@ export class Pokemon {
+
   subtractHP(pointsToSubtract: number): void {
-    const newHP = this._currentHP - pointsToSubtract;
+    const newHP: number = this._currentHP - pointsToSubtract;
     if (newHP < 0) {
       this._currentHP = 0;
+    } else {
+      this._currentHP = newHP;
     }
```
shorthand
```suggestion
    const newHP: number = this._currentHP - pointsToSubtract;
    this._currentHP = Math.max(0, newHP)
```

lukaszdutka commented on src/__tests__/fightClass.spec.ts:
```diff
@@ -1,14 +1,16 @@
-
+import { PokemonType } from "../pokemonClass"
+import { PokemonFactory } from "../pokemonFactory"
+import * as data from "../../assets/poke_data.json"
```
What I am afraid of is that: If we change `poke_data` for balance, then **tests will fail**. Any change in attack, hp or something of rattata, abra or any pokemon used there will result in change of expected numbers.
What I suggest is using constructor for pokemons: `new Pokemon('Poke1', '', 50, 50, 50)` - or something

AgataLudwiczynska committed:
Merge branch 'master' into MNG-89---Clicking-Start-should-render-ChoosePokemonScreen

AgataLudwiczynska committed:
Appendig function to the whole index.ts

mariusz-sm committed:
MNG-88_StartButtonActivation (#12)

* Add keyup to input

* Add activateStart.ts

* Uncomment html

* Add class to activate/disable start button

* Add activateStartButton function

* Change class name to button-enabled

AgataLudwiczynska committed:
Merge branch 'master' into MNG-89---Clicking-Start-should-render-ChoosePokemonScreen

AgataLudwiczynska committed:
Corrections

AgataLudwiczynska committed:
MNG-89   clicking start should render choose pokemon screen (#17)

* Function RenderChoosePokemonScreen

* Appendig function to the whole index.ts

* Corrections

AleksandraCyp committed:
MNG-95 Add GameHandler class (#14)

* Add GameHandler class

* Fix GameHandler class and add tests

* Fix tests

daria305 committed:
MNG-107 Add geodude instead of tentacruel

Filter moves to 4  for each pokemon
Use mostly pokemon type moves

AleksandraCyp committed:
Merge pull request #18 from lukaszdutka/MNG-107-Poke-data-corrections

MNG-107 Add geodude instead of tentacruel

AleksandraCyp committed:
MNG-103 choose pokemon system (#19)

* Add choosePokemonsPage.ts

* Add all functions for choose pokemon page

* Change event.target, loading time

AleksandraCyp committed:
MNG-94 MNG-108 Start a new game, add functions that will render the correct view (#20)

memeraki committed:
MNG-99 & MNG-100

> created menu and back button with options.
Options need to be fixed (now are default).
Animation also needs to look better

memeraki committed:
MNG-109 added

AgataLudwiczynska committed:
Blocking the start with unwritten name

memeraki committed:
added magicFunction

memeraki committed:
added animations for buttons

AleksandraCyp committed:
Merge pull request #21 from lukaszdutka/MNG-99&MNG-100-modals-for-action-buttons

MNG-99 & MNG-100

AgataLudwiczynska committed:
Style button

AgataLudwiczynska committed:
Merge branch 'master' into MNG-89---Clicking-Start-should-render-ChoosePokemonScreen

AgataLudwiczynska committed:
Blocking button when the name is too long

AleksandraCyp committed:
Merge pull request #23 from lukaszdutka/MNG-89---Clicking-Start-should-render-ChoosePokemonScreen

Mng 89   clicking start should render choose pokemon screen

AgataLudwiczynska committed:
MNG-98 Add useMango to gameHandler class (#22)

* Add useMango to gameHandler class

* corrections type of mango value

* Corrections

* useMango i PlayerClass

Co-authored-by: AleksandraCyp <73715885+AleksandraCyp@users.noreply.github.com>

AleksandraCyp committed:
MNG-97 MNG-101 change player&change pokemon (#24)

* Add switch pokemon and switch Player game handler methods

* Implement mango, change the structure (build up magic function)

* remove an unnecessary comment

daria305 committed:
Update test parametrization and do not use json data

daria305 committed:
Update test parametrization and do not use json data

daria305 committed:
Apply max shortcut

daria305 committed:
Merge branch 'MNG-102-finish-fight-implementation'

lukaszdutka reviewed (approved):
Overwhelming 🙃  Looks good :P
